Add dynamic features

DrNochi opened the pull request:
- In-memory db (Should probably be changed to persistent db later, bu in-memory ensures a fixed state after a restart for testing purposes at the moment)
- Basic flashcard adding (~~no categories yet~~)
- Basic display of categories (~~only name as of yet~~)

DrNochi committed:
Starting with the static prototype, add the first dynamic features to the app (basic flashcard adding & display of categories) [WIP]

Maaxxs committed:
Change tagline to something more useful

Maaxxs committed:
Display tagline and description on the dashboard

Maaxxs committed:
Add debugging output to the console and exception stack tracing

Maaxxs commented:
How to handle empty database values?
If no value provided as [here](https://github.com/manuel-tiedtke/M2N-Webanwendungen-2/blob/babe8eec607ff64738505c2a6d5a2e891ae98fc8/src/main/java/org/m2n/webapplications2/database/helpers/DbInit.java#L55), then it will display `undefined` in the Dashboard.

[![https://imgur.com/hatdO4w.png](https://imgur.com/hatdO4w.png)](https://imgur.com/hatdO4w.png)

So
1. Don't allow empty values in the database in the first place?
2. Catch that in [here](https://github.com/manuel-tiedtke/M2N-Webanwendungen-2/blob/1ac37f154f39719348842c553cd35faf1f684904/src/main/java/org/m2n/webapplications2/database/helpers/DbCategory.java#L42)

DrNochi commented on src/main/java/org/m2n/webapplications2/database/helpers/DbInit.java:
```diff
@@ -56,10 +56,10 @@ public static void insertTestData() throws DatabaseException {
+                "('Python', 'Programming Language', 'Python is awesome'),\n" +
+                "('Food', 'Maultaschen', 'Eat! Or you''ll starve!'),\n" +
+                "('Health', 'Sports', 'See category Food to stay healthy'),\n" +
+                "('Hardware', 'Security', 'Hardware Hacking Topics');\n");
```
Please also change the corresponding `.sql` file (should be `/src/main/sql/insert-test.sql` or something like that)

Maaxxs replied:
Done. 
429d830

DrNochi replied:
Thanks 👍

Maaxxs committed:
Change tagline to the same values as in
DbInit  babe8ee

DrNochi commented:
> How to handle empty database values?
> 1. Don't allow empty values in the database in the first place?
> 2. Catch that in [here](https://github.com/manuel-tiedtke/M2N-Webanwendungen-2/blob/1ac37f154f39719348842c553cd35faf1f684904/src/main/java/org/m2n/webapplications2/database/helpers/DbCategory.java#L42)

1. It should be allowed to not have a description or taglines
2. This "bug" should probably be handles in the `dashboard.js` file, as it is there that the value undefined arises. E.g. if the `description` is `null` in Java Jersey probably does not serialize this attribute. If JavaScript now tries to access the `description` in the deserialized version, it reads undefined.

Possible solution like:
`template.querySelector('h6').innerText = category.tagline ? category.tagline : ''`

DrNochi committed:
Make code style more consistent

Maaxxs commented:
> > How to handle empty database values?
> > 
> > 1. Don't allow empty values in the database in the first place?
> > 2. Catch that in [here](https://github.com/manuel-tiedtke/M2N-Webanwendungen-2/blob/1ac37f154f39719348842c553cd35faf1f684904/src/main/java/org/m2n/webapplications2/database/helpers/DbCategory.java#L42)
> 
>     1. It should be allowed to not have a description or taglines
> 
>     2. This "bug" should probably be handles in the `dashboard.js` file, as it is there that the value undefined arises. E.g. if the `description` is `null` in Java Jersey probably does not serialize this attribute. If JavaScript now tries to access the `description` in the deserialized version, it reads undefined.
> 
> 
> Possible solution like:
> `template.querySelector('h6').innerText = category.tagline ? category.tagline : ''`

I like that.

Maaxxs committed:
Don't display undefined for tagline or description

Resolve [this](https://github.com/manuel-tiedtke/M2N-Webanwendungen-2/pull/7#issuecomment-503003103)

Maaxxs committed:
Add more REST endpoints

DrNochi committed:
Make category editable

Maaxxs committed:
Display flashcards per category in "show cards"

DrNochi committed:
Validate category on update & prevent 'undefined' from showing on editing categories

DrNochi committed:
Repaire textarea-expander script

DrNochi committed:
Implement categories/tags when adding flashcards

DrNochi committed:
Use n:m-relationship endpoints to get flashcards in a category
